FOCIL (#2)
terencechain/consensus-specs: specs/_features/focil/engine-api.md

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Constants

| Name | Value |
| - | - |
| `INCLUSION_LIST_MAX_GAS` |  `uint64(4194304) = 2**22` |

### PayloadStatusV2

This structure contains the result of processing a payload. The fields are encoded as follows:

- `status`: `enum` - `"VALID" | "INVALID" | "SYNCING" | "ACCEPTED" | "INVALID_BLOCK_HASH" | "INVALID_INCLUSION_LIST"`
- `latestValidHash`: `DATA|null`, 32 Bytes - the hash of the most recent *valid* block in the branch defined by payload and its ancestors
- `validationError`: `String|null` - a message providing additional details on the validation error if the payload is classified as `INVALID` or `INVALID_BLOCK_HASH`.

## Methods

### engine_newPayloadV5

The request of this method is updated with [`ExecutionPayloadV5`](#ExecutionPayloadV5).

#### Request

* method: `engine_newPayloadV5`
* params:
  1. `executionPayload`: [`ExecutionPayloadV4`](#ExecutionPayloadV4).
  2. `expectedBlobVersionedHashes`: `Array of DATA`, 32 Bytes - Array of expected blob versioned hashes to validate.
  3. `parentBeaconBlockRoot`: `DATA`, 32 Bytes - Root of the parent beacon block.

>   4. `inclusionListTransactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)

Reviewer: ```suggestion
  4. `inclusionListTransactions`: `Array of DATA | null` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718), or null.
```

I think we should only pass an IL when necessary, e.g. when preparing to attest to a block from the current slot. When receiving an old beacon block, we should not care about ILs, and we shouldn't even have any to send (we can discard ILs as soon as their slot is over)

#### Response

* result: [`PayloadStatusV2`](#PayloadStatusV2)

#### Specification

1. Client software **MUST** respond to this method call in the following way:

>   * `{status: INVALID_INCLUSION_LISTING, latestValidHash: null, validationError: null}` if there are any leftover `inclusionListTransactions` that are not part of the `executionPayload`, they can be processed at the end of the `executionPayload`.

Reviewer 2: ```suggestion
  * `{status: INVALID_INCLUSION_LIST, latestValidHash: null, validationError: null}` if there are any leftover `inclusionListTransactions` that are not part of the `executionPayload`, they can be processed at the end of the `executionPayload`.
```

Reviewer 2: Do we need to mention the condition that execution payload has leftover gas?

> 1. Client software **MUST** respond to this method call in the following way:
>   * `{status: INVALID_INCLUSION_LIST, latestValidHash: null, validationError: null}` if there are any leftover `inclusionListTransactions` that are not part of the `executionPayload`, they can be processed at the end of the `executionPayload` or the block is full.

Reviewer: ```suggestion
1. Client software **MUST** respond to this method call in the following way:
  * `{status: INVALID_INCLUSION_LIST, latestValidHash: null, validationError: null}` if there are any leftover `inclusionListTransactions` that are not part of the `executionPayload`, but could still be included in it, meaning that they are valid against the post-state of `executionPayload` and their `gasLimit` is less than the remaining gas.
```

Reviewer: If the payload is otherwise valid, we should still get that information, because we want to know that we should not attest to it, but we don't  want to consider the payload invalid. Maybe change the name to `VALID_PAYLOAD_INVALID_INCLUSION_LIST` for clarity?

# Engine API -- FOCIL

Engine API changes introduced in FOCIL.

This specification is based on and extends [Engine API - Prague](./prague.md) specification.

Warning: This file should be placed in https://github.com/ethereum/execution-apis but I'm placing it here for convenience of reviewing.

> ## Table of contents

Reviewer 3: Should probably be LocalInclusionListSummary here too (throughout the whole section)

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Constants](#constants)
- [Structures](#structures)
  - [InclusionListSummaryV1](#inclusionlistsummaryv1)
  - [InclusionListSummaryListV1](#inclusionlistsummarylistv1)
  - [InclusionListStatusV1](#inclusionliststatusv1)
- [Methods](#methods)
  - [engine_newPayloadV5](#engine_newpayloadv5)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)
  - [engine_newInclusionListV1](#engine_newinclusionlistv1)
    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)
  - [engine_getInclusionListV1](#engine_getinclusionlistv1)
    - [Request](#request-2)
    - [Response](#response-2)
    - [Specification](#specification-2)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Constants

| Name | Value |
| - | - |

> | `INCLUSION_LIST_MAX_GAS` |  `uint64(4194304) = 2**22` |

Reviewer 3: `LOCAL_INCLUSION_LIST_MAX_GAS`
